bigquery/storage_integration_test.go

```diff
@@ -0,0 +1,370 @@
+	orderedQ.forceStorageAPI = true
+	it, err = orderedQ.Read(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	i = 0
+	for {
+		var s S
+		err := it.Next(&s)
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			t.Fatalf("failed to fetch via storage API: %v", err)
+		}
+		i++
+	}
+	t.Logf("%d lines read", i)
+	bqSession = it.arrowIterator.session.bqSession
+	if len(bqSession.Streams) > 1 {
+		t.Fatalf("should use just one stream as is ordered, but found %d", len(bqSession.Streams))
+	}
+	if i != it.TotalRows {
+		t.Fatalf("should have read %d rows, but read %d", it.TotalRows, i)
+	}
+	t.Logf("number of parallel streams for query `%s`: %d", orderedQ.Q, len(bqSession.Streams))
+	t.Logf("time taken: %d ms", time.Now().Sub(start).Milliseconds())
+}
+
+func TestIntegration_StorageReadQueryMorePages(t *testing.T) {
+	if client == nil {
+		t.Skip("Integration tests skipped")
+	}
+	ctx := context.Background()
+	start := time.Now()
+	table := "`bigquery-public-data.samples.github_timeline`"
+	sql := fmt.Sprintf(`SELECT repository_url as url, repository_owner as owner, repository_forks as forks FROM %s`, table)
+	// Don't forceStorageAPI usage and still see internally Storage API is selected
+	q := storageOptimizedClient.Query(sql)
+	it, err := q.Read(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !it.IsAccelerated() {
+		t.Fatal("expected query to use Storage API")
+	}
+
+	type S struct {
+		URL   NullString
+		Owner NullString
+		Forks NullInt64
+	}
+
+	var i uint64
+	for {
+		var s S
+		err := it.Next(&s)
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			t.Fatalf("failed to fetch via storage API: %v", err)
+		}
+		i++
+	}
+	t.Logf("%d lines read", i)
+	bqSession := it.arrowIterator.session.bqSession
+	if len(bqSession.Streams) == 0 {
+		t.Fatalf("should use more than one stream but found %d", len(bqSession.Streams))
+	}
+	if i != it.TotalRows {
+		t.Fatalf("should have read %d rows, but read %d", it.TotalRows, i)
+	}
+	t.Logf("number of parallel streams for query `%s`: %d", q.Q, len(bqSession.Streams))
+	t.Logf("bytes scanned for query `%s`: %d", q.Q, len(bqSession.Streams))
+	t.Logf("time taken: %d ms", time.Now().Sub(start).Milliseconds())
+}
+
+func TestIntegration_StorageReadCancel(t *testing.T) {
+	if client == nil {
+		t.Skip("Integration tests skipped")
+	}
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	table := "`bigquery-public-data.samples.github_timeline`"
+	sql := fmt.Sprintf(`SELECT repository_url as url, repository_owner as owner, repository_forks as forks FROM %s`, table)
+	q := storageOptimizedClient.Query(sql)
+	q.forceStorageAPI = true
+	it, err := q.Read(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !it.IsAccelerated() {
+		t.Fatal("expected query to use Storage API")
+	}
+
+	for {
+		var dst []Value
+		err := it.Next(&dst)
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				continue
```

Comment:
This seems odd, but I may be missing something here.  You keep looping if context expires?